### PR TITLE
improve test stability

### PR DESCRIPTION
**Pull request by frank-dong-ms**
1. add datatime when start/finish tests
2. add timeout for thread waiting to prevent infinite wait
3. fix cancel token cancel twice issue

### CODE REVIEW

File: src/Microsoft.ML.Core/Utilities/OrderedWaiter.cs [hunk 1]
```diff
@@ -90,7 +90,7 @@ public void Wait(long position, CancellationToken token = default(CancellationTo
                 ev = new WaitStats(position);
                 _waiters.Add(ev);
             }
-            ev.Event.Wait(token);
+            ev.Event.Wait(5*60*1000, token);
             if (_ex != null)
```
Comment (harishsk) [inline src/Microsoft.ML.Core/Utilities/OrderedWaiter.cs]: This is a five minute wait. Why five minutes? Can you please give more background behind this choice?

Comment (harishsk) [inline src/Microsoft.ML.Core/Utilities/OrderedWaiter.cs]: Can this be a constant that will apply to all use cases? Or does this need to be a parameter to the function? If it is a constant, can you please give it a name and use that instead?

File: src/Microsoft.ML.Data/Data/DataViewUtils.cs [hunk 2]
```diff
@@ -402,7 +402,7 @@ private static DataViewRowCursor ConsolidateCore(IChannelProvider provider, Data
                                             // The waiter event should never be null since this is only
                                             // called after a point where waiter.Register has been called.
                                             ch.AssertValue(waiterEvent);
-                                        waiterEvent.Wait();
+                                        waiterEvent.Wait(5*60*1000);
                                         waiterEvent = null;
```
Comment (harishsk) [inline src/Microsoft.ML.Data/Data/DataViewUtils.cs]: Similar comment as the one in OrderedWaiter.cs. If the same constant applies everywhere we should use a name.

File: src/Microsoft.Extensions.ML/ModelReloadToken.cs [hunk 3]
```diff
@@ -44,6 +44,12 @@ public ModelReloadToken()
         /// <summary>
         /// Used to trigger the change token when a reload occurs.
         /// </summary>
-        public void OnReload() => _cts.Cancel();
+        public void OnReload()
+        {
+            if (!_cts.IsCancellationRequested)
+                _cts.Cancel();
```
Comment (harishsk) [inline src/Microsoft.Extensions.ML/ModelReloadToken.cs]: The code seems to be saying if cancellation is not requested, call Cancel(). Am I reading that correctly? Can you please describe the root cause and the fix a bit more thoroughly in the pull request?

File: src/Microsoft.Extensions.ML/ModelReloadToken.cs [hunk 4]
```diff
@@ -44,6 +44,12 @@ public ModelReloadToken()
...
+            else
+                Console.WriteLine("ModelReloadToken has already been Canceled.");
+        }
```
Comment (harishsk) [inline src/Microsoft.Extensions.ML/ModelReloadToken.cs]: ML.NET is a library that can be used in environments where there is no console. I don't think we should be using Console.WriteLine here. 
What is the desired behavior? Is this for information only to the user or is this an error condition?

File: src/Microsoft.Extensions.ML/ModelReloadToken.cs (hunk 3 above)
Comment (frank-dong-ms) [reply src/Microsoft.Extensions.ML/ModelReloadToken.cs]: This means if the cancellation has not already been requested, then cancel the token.
we should always add this check as calling cancel twice will crash the process. This is one of the reason that test process is crashing.

---
In reply to: [350532410](https://github.com/dotnet/machinelearning/pull/4503#discussion_r350532410) [](ancestors = 350532410)

File: src/Microsoft.Extensions.ML/ModelReloadToken.cs (hunk 4 above)
Comment (frank-dong-ms) [reply src/Microsoft.Extensions.ML/ModelReloadToken.cs]: This is just to output some info into test command, this is not for user but rather a error info, I just output some info to see whether this is common test crash case.

---
In reply to: [350532971](https://github.com/dotnet/machinelearning/pull/4503#discussion_r350532971) [](ancestors = 350532971)

File: src/Microsoft.ML.Core/Utilities/OrderedWaiter.cs (hunk 1 above)
Comment (frank-dong-ms) [reply src/Microsoft.ML.Core/Utilities/OrderedWaiter.cs]: This caused most of our test hanging according to my local test result.
Some processor is waiting for signal to start but seems in some case it never can get the signal to run and the wait hangs the test process. This happens in multiple test cases so I just putting a pretty large timeout here, maybe we need discuss more to understand the root cause

---
In reply to: [350531489](https://github.com/dotnet/machinelearning/pull/4503#discussion_r350531489) [](ancestors = 350531489)

Comment (frank-dong-ms) [reply src/Microsoft.ML.Core/Utilities/OrderedWaiter.cs]: Yes, this should, this number is not reliable now so just put the number directly, let's discuss more in person.

---
In reply to: [350531730](https://github.com/dotnet/machinelearning/pull/4503#discussion_r350531730) [](ancestors = 350531730)

File: src/Microsoft.Extensions.ML/ModelReloadToken.cs (hunk 3 above)
Comment (eerhardt) [reply src/Microsoft.Extensions.ML/ModelReloadToken.cs]: > This is one of the reason that test process is crashing.

Can you provide a test log that shows this is where the test process is crashing? I don't see `.Cancel()` causing a process crash.

I made a console app that calls Cancel multiple times, and it doesn't even throw an exception.

File: src/Microsoft.Extensions.ML/ModelReloadToken.cs (hunk 4 above)
Comment (eerhardt) [reply src/Microsoft.Extensions.ML/ModelReloadToken.cs]: We should not be writing to the console inside of our libraries.

File: src/Microsoft.Extensions.ML/ModelReloadToken.cs (hunk 3 above)
Comment (frank-dong-ms) [reply src/Microsoft.Extensions.ML/ModelReloadToken.cs]: Below is the call stack I have reproduced in my local env:

The active test run was aborted. Reason: Test host process crashed : Unhandled Exception: System.AggregateException: One or more errors occurred. (Safe handle has been closed) ---> System.ObjectDisposedException: Safe handle has been closed
   at System.Runtime.InteropServices.SafeHandle.DangerousAddRef(Boolean& success)
   at System.StubHelpers.StubHelpers.SafeHandleAddRef(SafeHandle pHandle, Boolean& success)
   at Microsoft.Win32.Win32Native.SetEvent(SafeWaitHandle handle)
   at System.Threading.EventWaitHandle.Set()
   at Microsoft.Extensions.ML.FileLoaderTests.<>c__DisplayClass2_0.<can_reload_model>b__1() in C:\work\machinelearning\test\Microsoft.Extensions.ML.Tests\FileLoaderTests.cs:line 65
   at Microsoft.Extensions.Primitives.ChangeToken.<>c__DisplayClass0_0.<OnChange>b__0(Object s)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.CancellationTokenSource.ExecuteCallbackHandlers(Boolean throwOnFirstException)
   --- End of inner exception stack trace ---
   at System.Threading.CancellationTokenSource.ExecuteCallbackHandlers(Boolean throwOnFirstException)
   at Microsoft.Extensions.ML.ModelReloadToken.OnReload() in C:\work\machinelearning\src\Microsoft.Extensions.ML\ModelReloadToken.cs:line 47
   at Microsoft.Extensions.ML.FileModelLoader.WatcherChanged(Object sender, FileSystemEventArgs e) in C:\work\machinelearning\src\Microsoft.Extensions.ML\ModelLoaders\FileModelLoader.cs:line 78
   at System.IO.FileSystemWatcher.NotifyFileSystemEventArgs(WatcherChangeTypes changeType, String name)
   at System.IO.FileSystemWatcher.ParseEventBufferAndNotifyForEach(Byte[] buffer)
   at System.IO.FileSystemWatcher.ReadDirectoryChangesCallback(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* overlappedPointer)
   at System.Threading._IOCompletionCallback.PerformIOCompletionCallback(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* pOVERLAP)


---
In reply to: [350906884](https://github.com/dotnet/machinelearning/pull/4503#discussion_r350906884) [](ancestors = 350906884)

Comment (sharwell) [reply src/Microsoft.Extensions.ML/ModelReloadToken.cs]: > we should always add this check as calling cancel twice will crash the process. This is one of the reason that test process is crashing.

This is not correct. Cancellation a second time is a NOP, and `Cancel` itself is guarded by a check of `IsCancellationRequested`.

File: src/Microsoft.Extensions.ML/ModelReloadToken.cs [hunk 5]
```diff
@@ -44,6 +44,12 @@ public ModelReloadToken()
         /// <summary>
         /// Used to trigger the change token when a reload occurs.
         /// </summary>
-        public void OnReload() => _cts.Cancel();
+        public void OnReload()
+        {
+            if (!_cts.IsCancellationRequested)
```
Comment (sharwell) [inline src/Microsoft.Extensions.ML/ModelReloadToken.cs]: 📝 The changes in this file should be reverted. It duplicates a cancellation check that appears here:

https://github.com/dotnet/runtime/blob/b2073bd41e7ba0606000e5950e3d960d1d2eb36d/src/libraries/System.Private.CoreLib/src/System/Threading/CancellationTokenSource.cs#L582

File: src/Microsoft.ML.Core/Utilities/OrderedWaiter.cs [hunk 6]
```diff
@@ -90,7 +90,7 @@ public void Wait(long position, CancellationToken token = default(CancellationTo
                 ev = new WaitStats(position);
                 _waiters.Add(ev);
             }
-            ev.Event.Wait(token);
+            ev.Event.Wait(5*60*1000, token);
```
Comment (sharwell) [inline src/Microsoft.ML.Core/Utilities/OrderedWaiter.cs]: 📝 This change should be reverted. This is production code, not test code. If you want to add a delay, update the tests to pass a `CancellationToken` that will cancel after a period of time.

File: src/Microsoft.ML.Data/Data/DataViewUtils.cs [hunk 7]
```diff
@@ -402,7 +402,7 @@ private static DataViewRowCursor ConsolidateCore(IChannelProvider provider, Data
                                             // The waiter event should never be null since this is only
                                             // called after a point where waiter.Register has been called.
                                             ch.AssertValue(waiterEvent);
-                                        waiterEvent.Wait();
+                                        waiterEvent.Wait(5*60*1000);
```
Comment (sharwell) [inline src/Microsoft.ML.Data/Data/DataViewUtils.cs]: 📝 This change should be reverted for the same reason I mentioned for **OrderedWaiter.cs**.

**Comment by frank-dong-ms** [comment]
Thanks all for information, close this PR now and working on new one

**Pull request closed without merging**